File: skills_manager/backup.py
"""Portable backup and restore for the managed skill system."""
from __future__ import annotations

import shutil
import time
from pathlib import Path
from typing import Any

from . import adapters, materializer, store


def backup_root(export_path: str | Path) -> Path:
    path = Path(export_path).expanduser().resolve()
    return path if path.name == "agent-skills-backup" else path / "agent-skills-backup"


def rendered_list(client: str) -> list[dict[str, Any]]:
    rendered_dir = adapters.adapter(client).global_dir()
    return list(materializer.actual_rendered(rendered_dir).values())


def dry_run_export(export_path: str | Path | None = None) -> dict[str, Any]:
    target = str(backup_root(export_path or "./agent-skills-backup"))
    return {
        "target": target,
        "include": [
            str(store.skills_root()),
            str(store.manifests_root()),
            str(store.transactions_root()),
            str(store.presets_root()),
            str(store.logs_root()),
            str(store.inbox_dir()),
        ],
        "rendered_metadata_only": {"claude": rendered_list("claude"), "codex": rendered_list("codex")},
    }


def copy_if_exists(src: Path, dst: Path) -> None:
    if not src.exists():
        return
    if src.is_dir():
        shutil.copytree(src, dst, dirs_exist_ok=True)
    else:
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dst)


def export(export_path: str | Path) -> dict[str, Any]:
    root = backup_root(export_path)
    root.mkdir(parents=True, exist_ok=True)
    copy_if_exists(store.skills_root(), root / "skills-store" / "skills")
    copy_if_exists(store.manifests_root(), root / "skills-store" / "manifests")
    copy_if_exists(store.transactions_root(), root / "skills-store" / "transactions")
    copy_if_exists(store.presets_root(), root / "skills-store" / "presets")
    copy_if_exists(store.logs_root(), root / "skills-store" / "logs")
    copy_if_exists(store.inbox_dir(), root / "inbox" / "agents-skills")
    store.write_json(root / "rendered" / "claude-skills-list.json", rendered_list("claude"))
    store.write_json(root / "rendered" / "codex-skills-list.json", rendered_list("codex"))
    manifest = {"version": store.VERSION, "created_at": time.time(), "kind": "agent-skills-backup"}
    store.write_json(root / "manifest.json", manifest)
    return {"ok": True, "backup": str(root), "manifest": manifest}


def normalize_backup(path: str | Path) -> Path:
    root = Path(path).expanduser().resolve()
    if (root / "manifest.json").exists():
        return root
    if (root / "agent-skills-backup" / "manifest.json").exists():
        return root / "agent-skills-backup"
    return root


def restore_plan(path: str | Path) -> dict[str, Any]:
    root = normalize_backup(path)
    return {
        "backup": str(root),
        "exists": (root / "manifest.json").exists(),
        "copies": [
            {"from": str(root / "skills-store" / "skills"), "to": str(store.skills_root())},
            {"from": str(root / "skills-store" / "manifests"), "to": str(store.manifests_root())},
            {"from": str(root / "skills-store" / "transactions"), "to": str(store.transactions_root())},
            {"from": str(root / "skills-store" / "presets"), "to": str(store.presets_root())},
            {"from": str(root / "skills-store" / "logs"), "to": str(store.logs_root())},
            {"from": str(root / "inbox" / "agents-skills"), "to": str(store.inbox_dir())},
        ],
        "after": ["skills-manager materialize --client all", "skills-manager doctor"],
    }


def restore(path: str | Path, dry_run: bool = True) -> dict[str, Any]:
    plan = restore_plan(path)
    if not plan["exists"]:
        return {"ok": False, "error": f"backup manifest not found: {plan['backup']}", "plan": plan}
    if dry_run:
        return {"ok": True, "dry_run": True, "plan": plan}
    for copy in plan["copies"]:
        copy_if_exists(Path(copy["from"]), Path(copy["to"]))
    return {"ok": True, "dry_run": False, "plan": plan, "message": "restore copied store/inbox data; run materialize next"}

File: src/core/backup.ts
import { cp, mkdir, stat } from "node:fs/promises";
import { basename, dirname, join, resolve } from "node:path";
import { actualRendered } from "./materializer.js";
import { adapter } from "./adapters.js";
import { expandUser, inboxDir, logsRoot, manifestsRoot, presetsRoot, skillsRoot, storeRoot, transactionsRoot, type Env } from "./paths.js";
import { shouldSkipLocalJunkPath, VERSION, writeJson } from "./store.js";

export function backupRoot(exportPath: string, env: Env = process.env): string {
  const path = resolve(expandUser(exportPath, env));
  return basename(path) === "agent-skills-backup" ? path : join(path, "agent-skills-backup");
}

export function preMigrationBackupRoot(exportPath: string, env: Env = process.env): string {
  const path = resolve(expandUser(exportPath, env));
  return basename(path) === "agent-skills-pre-migration-backup" ? path : join(path, "agent-skills-pre-migration-backup");
}

async function exists(path: string): Promise<boolean> {
  try {
    await stat(path);
    return true;
  } catch {
    return false;
  }
}

async function renderedList(client: "claude" | "codex", env: Env = process.env): Promise<unknown[]> {
  return Object.values(await actualRendered(adapter(client, env).globalDir(), env));
}

export async function dryRunExport(exportPath = "./agent-skills-backup", env: Env = process.env): Promise<Record<string, unknown>> {
  return {
    target: backupRoot(exportPath, env),
    kind: "agent-skills-backup",
    note: "Managed-state backup: rendered Claude/Codex skill directories are metadata-only, not raw-copied.",
    include: [skillsRoot(env), manifestsRoot(env), transactionsRoot(env), presetsRoot(env), logsRoot(env), inboxDir(env)],
    rendered_metadata_only: {
      claude: await renderedList("claude", env),
      codex: await renderedList("codex", env)
    }
  };
}

type RawCopy = {
  name: "claude" | "codex" | "agents";
  from: string;
  to: string;
  exists: boolean;
};

async function rawPreMigrationCopies(root: string, env: Env = process.env): Promise<RawCopy[]> {
  const copies = [
    { name: "claude" as const, from: adapter("claude", env).globalDir(), to: join(root, "raw", "claude-skills") },
    { name: "codex" as const, from: adapter("codex", env).globalDir(), to: join(root, "raw", "codex-skills") },
    { name: "agents" as const, from: inboxDir(env), to: join(root, "raw", "agents-skills") }
  ];
  return Promise.all(copies.map(async (copy) => ({ ...copy, exists: await exists(copy.from) })));
}

export async function dryRunPreMigrationBackup(exportPath = "./agent-skills-pre-migration-backup", env: Env = process.env): Promise<Record<string, unknown>> {
  const target = preMigrationBackupRoot(exportPath, env);
  return {
    target,
    kind: "agent-skills-pre-migration-backup",
    note: "Raw pre-migration backup: full copies of rendered Claude/Codex skill dirs and the ~/.agents/skills inbox. This is intentionally separate from managed-state backup/restore.",
    raw_copies: await rawPreMigrationCopies(target, env)
  };
}

export async function copyIfExists(src: string, dst: string): Promise<void> {
  if (!(await exists(src))) {
    return;
  }
  await mkdir(dirname(dst), { recursive: true });
  await cp(src, dst, {
    recursive: true,
    force: true,
    filter: async (source) => !shouldSkipLocalJunkPath(source, { root: src, allowManagedDotRoot: true })
  });
}

export async function exportBackup(exportPath: string, env: Env = process.env): Promise<Record<string, unknown>> {
  const root = backupRoot(exportPath, env);
  await mkdir(root, { recursive: true });
  await copyIfExists(skillsRoot(env), join(root, "skills-store", "skills"));
  await copyIfExists(manifestsRoot(env), join(root, "skills-store", "manifests"));
  await copyIfExists(transactionsRoot(env), join(root, "skills-store", "transactions"));
  await copyIfExists(presetsRoot(env), join(root, "skills-store", "presets"));
  await copyIfExists(logsRoot(env), join(root, "skills-store", "logs"));
  await copyIfExists(inboxDir(env), join(root, "inbox", "agents-skills"));
  await writeJson(join(root, "rendered", "claude-skills-list.json"), await renderedList("claude", env));
  await writeJson(join(root, "rendered", "codex-skills-list.json"), await renderedList("codex", env));
  const manifest = { version: VERSION, created_at: Date.now() / 1000, kind: "agent-skills-backup" };
  await writeJson(join(root, "manifest.json"), manifest);
  return { ok: true, backup: root, manifest };
}

export async function exportPreMigrationBackup(exportPath: string, env: Env = process.env): Promise<Record<string, unknown>> {
  const root = preMigrationBackupRoot(exportPath, env);
  await mkdir(root, { recursive: true });
  const rawCopies = await rawPreMigrationCopies(root, env);
  const copied: RawCopy[] = [];
  for (const copy of rawCopies) {
    if (!copy.exists) {
      continue;
    }
    await copyIfExists(copy.from, copy.to);
    copied.push(copy);
  }
  const manifest = {
    version: VERSION,
    created_at: Date.now() / 1000,
    kind: "agent-skills-pre-migration-backup",
    note: "Full raw copies for safety before migration. Not used by skills-manager restore.",
    raw_copies: rawCopies
  };
  await writeJson(join(root, "manifest.json"), manifest);
  return { ok: true, backup: root, manifest, copied };
}

export async function normalizeBackup(path: string, env: Env = process.env): Promise<string> {
  const root = resolve(expandUser(path, env));
  if (await exists(join(root, "manifest.json"))) {
    return root;
  }
  if (await exists(join(root, "agent-skills-backup", "manifest.json"))) {
    return join(root, "agent-skills-backup");
  }
  return root;
}

export async function restorePlan(path: string, env: Env = process.env): Promise<Record<string, unknown>> {
  const root = await normalizeBackup(path, env);
  return {
    backup: root,
    exists: await exists(join(root, "manifest.json")),
    copies: [
      { from: join(root, "skills-store", "skills"), to: skillsRoot(env) },
      { from: join(root, "skills-store", "manifests"), to: manifestsRoot(env) },
      { from: join(root, "skills-store", "transactions"), to: transactionsRoot(env) },
      { from: join(root, "skills-store", "presets"), to: presetsRoot(env) },
      { from: join(root, "skills-store", "logs"), to: logsRoot(env) },
      { from: join(root, "inbox", "agents-skills"), to: inboxDir(env) }
    ],
    after: ["skills-manager materialize --client all", "skills-manager doctor"]
  };
}

export async function restoreBackup(path: string, options: { dryRun?: boolean; env?: Env } = {}): Promise<Record<string, unknown>> {
  const env = options.env ?? process.env;
  const plan = await restorePlan(path, env);
  if (!plan.exists) {
    return { ok: false, error: `backup manifest not found: ${plan.backup}`, plan };
  }
  if (options.dryRun ?? true) {
    return { ok: true, dry_run: true, plan };
  }
  for (const item of plan.copies as Array<{ from: string; to: string }>) {
    await copyIfExists(item.from, item.to);
  }
  return { ok: true, dry_run: false, plan, message: "restore copied store/inbox data; run materialize next" };
}

export function currentStoreRoot(env: Env = process.env): string {
  return storeRoot(env);
}
